http_service/bugbug_http/utils.py

```diff
@@ -30,3 +35,44 @@ def get_hgmo_stack(branch: str, revision: str) -> list:
     r = requests.get(url)
     r.raise_for_status()
     return r.json()["changesets"]
+
+
+# A simple TTL cache to use with models. Because we expect the number of models
+# in the service to not be very large, simplicity of implementation is
+# preferred to algorithmic efficiency of operations.
+#
+# Called an 'Idle' TTL cache because TTL of items is reset after every get
+Key = TypeVar("Key")
+Value = TypeVar("Value")
+
+
+class IdleTTLCache(Generic[Key, Value]):
+    def __init__(self, ttl: timedelta):
+        self.ttl = ttl
+        self.items_last_touched: Dict[Key, datetime.datetime] = {}
+        self.items: Dict[Key, Value] = {}
+
+    def __setitem__(self, key, item):
+        self.items[key] = item
+        self.items_last_touched[key] = datetime.datetime.now()
+
+    def __getitem__(self, key):
+        item = self.items[key]
+        self.items_last_touched[key] = datetime.datetime.now()
+        return item
+
+    def __contains__(self, key):
+        return key in self.items
+
+    def purge_expired_entries(self):
+        purge_entries_before = datetime.datetime.now() - self.ttl
+        for (key, time_last_touched) in list(self.items_last_touched.items()):
+            if time_last_touched < purge_entries_before:
+                del self.items_last_touched[key]
+                del self.items[key]
+
+    def start_ttl_thread(self):
+        self.purge_expired_entries()
+        timer = threading.Timer(self.ttl.total_seconds(), self.start_ttl_thread)
+        timer.setDaemon(True)
+        timer.start()
```

Comment:
It'd probably be preferable to have a single thread with an infinite loop which waits for the TTL and then purges, rather than create a new thread for every period.

Reply:
Will do.

I think you're right - a single thread with a sleep seems like it'd run more efficiently as well as result in marginally more readable code.

I implemented this with a Timer assuming (without checking the assumption) that Timers give up the GIL while waiting to run, whereas threads do not.  Digging deeper after reading your comment, I do think Timers work this way. However, if I correctly understood this [article on the GIL by a core python developer](http://jessenoller.com/blog/2009/02/01/python-threads-and-the-global-interpreter-lock), so do Threads when they are executing a sleep.

Using a Thread with a sleep and an infinite loop therefore seems strictly superior to a Timer, since it looks like neither take up the GIL while waiting, but a Thread with a loop has less overhead than a Timer.